src/nemos/solver.py

```diff
@@ -0,0 +1,509 @@
+    solver_kwargs :
+        Additional keyword arguments to be passed to the solver during instantiation.
+    """
+
+    allowed_algorithms: List[str] = []
+
+    def __init__(
+        self,
+        solver_name: str,
+        solver_kwargs: Optional[dict] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self._check_solver(solver_name)
+        self._solver_name = solver_name
+        if solver_kwargs is None:
+            self._solver_kwargs = dict()
+        else:
+            self._solver_kwargs = solver_kwargs
+        self._check_solver_kwargs(self.solver_name, self.solver_kwargs)
+
+    @property
+    def solver_name(self):
+        return self._solver_name
+
+    @solver_name.setter
+    def solver_name(self, solver_name: str):
+        self._check_solver(solver_name)
+        self._solver_name = solver_name
+
+    @property
+    def solver_kwargs(self):
+        return self._solver_kwargs
+
+    @solver_kwargs.setter
+    def solver_kwargs(self, solver_kwargs: dict):
+        self._check_solver_kwargs(self.solver_name, solver_kwargs)
+        self._solver_kwargs = solver_kwargs
+
+    def _check_solver(self, solver_name: str):
+        """
+        Ensure the provided solver name is allowed.
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver to be checked.
+
+        Raises
+        ------
+        ValueError
+            If the provided solver name is not in the list of allowed optimizers.
+        """
+        if solver_name not in self.allowed_algorithms:
+            raise ValueError(
+                f"Solver `{solver_name}` not allowed for "
+                f"{self.__class__} regularization. "
+                f"Allowed solvers are {self.allowed_algorithms}."
+            )
+
+    @staticmethod
+    def _check_solver_kwargs(solver_name, solver_kwargs):
+        """
+        Check if provided solver keyword arguments are valid.
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver.
+        solver_kwargs :
+            Additional keyword arguments for the solver.
+
+        Raises
+        ------
+        NameError
+            If any of the solver keyword arguments are not valid.
+        """
+        solver_args = inspect.getfullargspec(getattr(jaxopt, solver_name)).args
+        undefined_kwargs = set(solver_kwargs.keys()).difference(solver_args)
+        if undefined_kwargs:
+            raise NameError(
+                f"kwargs {undefined_kwargs} in solver_kwargs not a kwarg for jaxopt.{solver_name}!"
+            )
+
+    def instantiate_solver(
+        self, loss: Callable, *args: Any, **kwargs: Any
+    ) -> SolverRunner:
+        """Abstract method to instantiate a solver with a given loss function."""
+        # check that the loss is Callable
+        utils.assert_is_callable(loss, "loss")
+        return self.get_runner(loss, *args, **kwargs)
+
+    def get_runner(
```

Comment:
copying from earlier comment:


looks like this always gets called from within `instantiate_solver`. do we want to just have it live in the superclass version of instantiate_solver and have the subclasses call `super().instantiate_solver()` at the bottom?

since this method never gets called by itself and anyone implementing their own solver should call it in `instantiate_solver()`